### DISCUSSION TITLE
How to create child actors with dependency injection

### DISCUSSION THREAD

**Discussion by Kosntantin** [akkadotnet/akka.net#5854] [General]
I want to use akka for concurrent message processing. I am working with objects called Fixture. My design is to have a supervisor actor that gets all the messages and then routes them to another actor that handles all messages for the specific Fixture. To do this I have a dictionary of <FixtureID,IActorRef> in the supervisor actor. Naturally, if the supervisor receives a message for a Fixture that currently has no IActorRef, the supervisor needs to create a child actor and add it the dictionary. This is where I run into trouble. I am unable to create a child actor because it cannot resolve the dependencies.

This is how I register the supervising actor so I can pass it to the class that will send him the messages:

```
services.AddSingleton<IActorRef>(sp =>
                {
                    var bootstrap = BootstrapSetup.Create();
                    var di = DependencyResolverSetup.Create(sp);
                    var actorSystemSetup = bootstrap.And(di);
                    var _actorSystem = ActorSystem.Create("AspNetDemo", actorSystemSetup);
                    var resolver = DependencyResolver.For(_actorSystem);
                    var props = resolver.Props<SupervisingActor>();

                    return _actorSystem.ActorOf(props, "supervisor");
                });
```
This is how I create child actors in the supervisor:
`
Context.ActorOf(Context.DI().Props<FixtureActor>(), $"FixtureActor{message.Id}");`
The supervisor actor is created successfully but when it tries to spawn a child actor it throws this error: 'The Dependency Resolver has not been configured yet'

My code is based on this: https://getakka.net/articles/actors/dependency-injection.html Note: I do not want to use another DI Container.

**Comment by Arkatufus** [comment]
@Kosntantin 
I'm assuming you're using ASP.NET? Which version is it?
Which version of Akka.NET are you using?
Is there a limitation that restricts you to use the deprecated `Akka.DI` package? can you switch to the new `Akka.DependencyInjection` package?
Which `Akka.DI` package are you using?

**Comment by Kosntantin** [comment]
It is .NET 6 and the Akka version is 1.4.37. It is a Worker service, so still a HostedService but not ASP. I am using the new Akka.DependencyInjection package to create the supervisor actor but I am not sure how to create the child actors using the package. This is why I used the depricated. I have nothing that restricts me to using the old pakcage. Can you show me how to use the new one to create child actors?

**Reply by Arkatufus** [reply]
I think you need to re-arrange everything, the way you're setting everything up is not quite safe.
Here's an example project that uses `Akka.Hosting` and `Akka.DependencyInjection` in a safe way, the `ActorSystem` is already wrapped inside a `IHostedService` and its life cycle is properly hooked into the `IHost` life cycle:

https://github.com/Arkatufus/Akka.Net.Samples/tree/master/Akka.Hosting/HostedService

**Reply by Arkatufus** [reply]
cc @Kosntantin

**Reply by valeriob** [reply]
Thanks for the link to the example, I'm having problems doing the same thing with a ShardedRegion, but I do not understand how to create an actor Props with both DI and passing the entityId
https://petabridge.com/blog/intro-akka-hosting/

![image](https://user-images.githubusercontent.com/1522516/176669055-60ff0a0b-8d6e-42fc-a3e5-7c41bc535dcb.png)


Thanks

EDIT: 
i was able to make it work via 
compositePropsFactory: (system, registry) => s => DependencyResolver.For(system).Props<IndexActorEntity>(s)
is it correct ?

**Reply by Aaronontheweb** [reply]
@valeriob that is correct!
